### ISSUE TITLE
Changing `DYLINT_RUSTFLAGS` does not cause lints to be re-run

### ISSUE THREAD

**Issue by dcastro** [trailofbits/dylint#1690]
Is there any way of setting the level of all lints to "Error" when running dylint? Like how one can run `cargo clippy -- -D warnings` to turn all warnings into errors.

I know they can be set individually when declaring the lint with `declare_late_lint`, but I'd like to define lint rule that are set to Warn by default, but turn everything into an Error on the CI.

I can't find an option in the CLI's --help for this.

**Comment by smoelius** [comment]
I think either of the following should work:
```sh
RUSTFLAGS='-D warnings' cargo dylint ...
```
```sh
DYLINT_RUSTFLAGS='-D warnings' cargo dylint ...
```
The advantage of the latter is that it does not cause all of the dependencies to be rebuilt.

**Comment by dcastro** [comment]
@smoelius thanks, `DYLINT_RUSTFLAGS` didn't work, but `RUSTFLAGS` did!

**Comment by smoelius** [comment]
What happened when you tried to use `DYLINT_RUSTFLAGS`?

**Comment by dcastro** [comment]
@smoelius same thing as if I hadn't set `DYLINT_RUSTFLAGS`. Those lints were still reported as warnings.

**Comment by smoelius** [comment]
Yeah, it looks there's a bug.

**Comment by dcastro** [comment]
@smoelius I'm pretty sure this was due to me using a custom linker, as you pointed out in #1691.

I tried this again with `DYLINT_RUSTFLAGS='-D warnings' RUSTFLAGS='' cargo dylint --all -- --workspace --all-targets`, and the warnings now turn into errors as expected.

Thanks again!

**Comment by smoelius** [comment]
Thanks for your note, @dcastro. But I think something more subtle is going on.

Ideally, this should work:
```sh
$ cargo dylint ...
# compilation produces warnings but succeeds
$ DYLINT_RUSTFLAGS='-D warnings' cargo dylint ...
# compilation fails
```

In particular, the latter step does not cause the lints to be re-run. It just reuses the results of the first run.

Now, if something else changes (e.g., a source file changes), the lints are re-run. But changing `DYLINT_RUSTFLAGS` ought to be enough. Currently, it seems it's not.

And, for future me, I think something in here needs to change: https://github.com/trailofbits/dylint/blob/a4d46127c83a48fc05cbc85a1638a7bdc3252711/driver/src/lib.rs#L251-L262

**Comment by smoelius** [comment]
> Ideally, this should work:
> ```sh
> $ cargo dylint ...
> # compilation produces warnings but succeeds
> $ DYLINT_RUSTFLAGS='-D warnings' cargo dylint ...
> # compilation fails
> ```

I'm no longer convinced this is true.

Note that `cargo dylint` can be run on any Cargo package. So, Cargo would have to be `DYLINT_RUSTFLAGS`-aware to know that a change to that environment variable would require workspace members to be rebuilt, but not other packages.

The code I pointed to in https://github.com/trailofbits/dylint/issues/1690#issuecomment-3089922259 tells rustc this, but not  Cargo. I'm not sure there's a way to tell Cargo this.

I thought Clippy had an environment variable that worked as I described, but I can't now find it.

I'm going to leave this open for now, but I think there may be nothing more I can do here.
